pkg/distributor/distributor.go

```diff
@@ -241,196 +241,206 @@ func (d *Distributor) stopping(_ error) error {
-		return resp, err
+		return nil, err
+	}
+	if err != nil {
+		return nil, err
 	}
-	return resp, err
+	return connect.NewResponse(new(pushv1.PushResponse)), err
 }
 
 func (d *Distributor) GetProfileLanguage(series *distributormodel.ProfileSeries) string {
 	if series.Language != "" {
 		return series.Language
 	}
-	if len(series.Samples) == 0 {
-		return "unknown"
-	}
 	lang := series.GetLanguage()
 	if lang == "" {
-		lang = pprof.GetLanguage(series.Samples[0].Profile)
+		lang = pprof.GetLanguage(series.Sample.Profile)
 	}
 	series.Language = lang
 	return series.Language
 }
 
-func (d *Distributor) PushParsed(ctx context.Context, req *distributormodel.PushRequest) (resp *connect.Response[pushv1.PushResponse], err error) {
-	now := model.Now()
+func (d *Distributor) PushBatch(ctx context.Context, req *distributormodel.PushRequest) error {
 	tenantID, err := tenant.ExtractTenantIDFromContext(ctx)
 	if err != nil {
-		return nil, connect.NewError(connect.CodeUnauthenticated, err)
+		return connect.NewError(connect.CodeUnauthenticated, err)
 	}
-
-	logger := log.With(d.logger, "tenant", tenantID)
-
-	req.TenantID = tenantID
-	for _, series := range req.Series {
-		serviceName := phlaremodel.Labels(series.Labels).Get(phlaremodel.LabelNameServiceName)
-		if serviceName == "" {
-			series.Labels = append(series.Labels, &typesv1.LabelPair{Name: phlaremodel.LabelNameServiceName, Value: phlaremodel.AttrServiceNameFallback})
-		}
-		sort.Sort(phlaremodel.Labels(series.Labels))
+	if len(req.Series) == 0 {
+		return noNewProfilesReceivedError()
 	}
 
-	haveRawPprof := req.RawProfileType == distributormodel.RawProfileTypePPROF
-	d.bytesReceivedTotalStats.Inc(int64(req.RawProfileSize))
-	d.bytesReceivedStats.Record(float64(req.RawProfileSize))
-	if !haveRawPprof {
+	d.bytesReceivedTotalStats.Inc(int64(req.ReceivedCompressedProfileSize))
+	d.bytesReceivedStats.Record(float64(req.ReceivedCompressedProfileSize))
+	if req.RawProfileType != distributormodel.RawProfileTypePPROF {
 		// if a single profile contains multiple profile types/names (e.g. jfr) then there is no such thing as
 		// compressed size per profile type as all profile types are compressed once together. So we can not count
 		// compressed bytes per profile type. Instead we count compressed bytes per profile.
 		profName := req.RawProfileType // use "jfr" as profile name
-		d.metrics.receivedCompressedBytes.WithLabelValues(string(profName), tenantID).Observe(float64(req.RawProfileSize))
+		d.metrics.receivedCompressedBytes.WithLabelValues(string(profName), tenantID).Observe(float64(req.ReceivedCompressedProfileSize))
+	}
+
+	results := make([]error, len(req.Series))
+	errorsMutex := new(sync.Mutex)
+	wg := new(sync.WaitGroup)
+	for i, s := range req.Series {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			itErr := util.RecoverPanic(func() error {
+				return d.pushSeries(ctx, s, req.RawProfileType, tenantID)
+			})
+			errorsMutex.Lock()
+			results[i] = itErr()
```

Comment:
This is a bug, this should not be invoked under lock

Reply:
https://github.com/grafana/pyroscope/pull/4342/commits/d615aef13117389ca18fd20e54c02d3aa7941da5